### PR TITLE
Calculation of the progress of the project

### PR DESCRIPTION
**Pull request by arek1-618**
Calculation of the progress of the project.

### CODE REVIEW

File: modules/Project/handlers/ProjectHandler.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	 * EntityAfterSave handler function.
+	 *
+	 * @param \App\EventHandler $eventHandler
+	 */
+	public function entityAfterSave(\App\EventHandler $eventHandler)
+	{
+		$recordModel = $eventHandler->getRecordModel();
+		if (!$recordModel->isNew()) {
+			$delta = $recordModel->getPreviousValue();
+			foreach ($delta as $name => $value) {
```
Comment (rskrzypczak) [inline modules/Project/handlers/ProjectHandler.php]: without foreach.
$recordModel->getPreviousValue('parentid') ! == false

File: modules/Project/handlers/ProjectHandler.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	/**
+	 * EntityChangeState handler function.
+	 *
+	 * @param \App\EventHandler $eventHandler
+	 */
+	public function entityChangeState(\App\EventHandler $eventHandler)
+	{
+		$recordModel = $eventHandler->getRecordModel();
+		if (!$recordModel->isEmpty('parentid')) {
+			Vtiger_Module_Model::getInstance('Project')->updateProgress($recordModel->get('parentid'));
```
Comment (rskrzypczak) [inline modules/Project/handlers/ProjectHandler.php]:
```suggestion
			$recordModel->getModule()->updateProgress($recordModel->get('parentid'));
```

File: modules/Project/models/Module.php [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
 
+/**
+ * Project module model class.
+ *
+ * @package   Model
+ *
+ * @copyright YetiForce Sp. z o.o
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Arkadiusz Adach <a.adach@yetiforce.com>
+ */
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: You can't change license

File: modules/Project/models/Module.php [hunk 4]
```diff
@@ -23,7 +23,112 @@ public function getSideBarLinks($linkParams)
...
+			}
+		}
+		$this->calculateProgressOfTasks($recordModel, $estimatedWorkTime, $progressInHours);
+		if ($estimatedWorkTime) {
+			$projectProgress = round((100 * $progressInHours) / $estimatedWorkTime);
+		} else {
+			$projectProgress = 0;
+		}
+		$recordModel = Vtiger_Record_Model::getInstanceById($id, $this->getName());
+		$recordModel->set('progress', $projectProgress . '%');
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: change type field to Percentage

File: modules/Project/models/Module.php [hunk 5]
```diff
@@ -23,7 +23,112 @@ public function getSideBarLinks($linkParams)
...
+	 * @throws \App\Exceptions\AppException
+	 *
+	 * @return array
+	 */
+	public function updateProgress(int $id, float $estimatedWorkTime = 0, float $progressInHours = 0, ?int $parentId = null): array
+	{
+		if (!App\Record::isExists($id)) {
+			return [];
+		}
+		$recordModel = Project_Record_Model::getInstanceById($id);
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: Vtiger_Record_Model

File: modules/Project/models/Module.php [hunk 6]
```diff
@@ -23,7 +23,112 @@ public function getSideBarLinks($linkParams)
...
+	 */
+	public function calculateProgressOfChildren(int $id, float $estimatedWorkTime = 0, float $progressInHours = 0)
+	{
+		if (!App\Record::isExists($id)) {
+			return [
+				'estimatedWorkTime' => 0,
+				'projectProgress' => 0
+			];
+		}
+		$recordModel = Project_Record_Model::getInstanceById($id);
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: Vtiger_Record_Model

File: modules/Project/models/Record.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * @author    Arkadiusz Adach <a.adach@yetiforce.com>
+ */
+class Project_Record_Model extends Vtiger_Record_Model
+{
+	/**
+	 * Get children by parent ID.
+	 *
+	 * @return \Generator
+	 */
+	public function getChildren()
```
Comment (rskrzypczak) [inline modules/Project/models/Record.php]: You really need record models return?

File: modules/ProjectTask/handlers/ProjectTaskHandler.php [hunk 8]
```diff
@@ -3,54 +3,72 @@
...
 	/**
 	 * EntityChangeState handler function.
 	 *
-	 * @param App\EventHandler $eventHandler
+	 * @param \App\EventHandler $eventHandler
 	 */
-	public function entityChangeState(App\EventHandler $eventHandler)
+	public function entityChangeState(\App\EventHandler $eventHandler)
 	{
 		Vtiger_Module_Model::getInstance('ProjectMilestone')->updateProgressMilestone($eventHandler->getRecordModel()->get('projectmilestoneid'));
```
Comment (rskrzypczak) [inline modules/ProjectTask/handlers/ProjectTaskHandler.php]: I think this is to remove.

File: modules/ProjectTask/handlers/ProjectTaskHandler.php [hunk 9]
```diff
@@ -3,53 +3,68 @@
...
+				} elseif ($name === 'projectid') {
+					$calculateProject[$recordModel->get($name)] = true;
+					$calculateProject[$value] = true;
 				}
 			}
+			$milestoneModel = Vtiger_Module_Model::getInstance('ProjectMilestone');
+			foreach ($calculateMilestone as $milestoneId => $val) {
+				$milestoneModel->updateProgressMilestone($milestoneId);
+			}
+			/*$projectModel = Vtiger_Module_Model::getInstance('Project');
```
Comment (rskrzypczak) [inline modules/ProjectTask/handlers/ProjectTaskHandler.php]: ?

File: modules/ProjectMilestone/models/Record.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+			yield self::getInstanceById($projectInfo[$instance->table_index]);
+		}
+	}
+
+	/**
+	 * Get progress.
+	 *
+	 * @return float
+	 */
+	public function getProgress(): float
```
Comment (rskrzypczak) [inline modules/ProjectMilestone/models/Record.php]: ?

File: modules/Project/models/Module.php [hunk 11]
```diff
@@ -23,7 +24,158 @@ public function getSideBarLinks($linkParams)
...
+	{
+		$relatedListView = Vtiger_RelationListView_Model::getInstance($recordModel, 'ProjectMilestone');
+		$relatedListView->getRelationModel()->set('QueryFields', [
+			'projectmilestone_progress' => 'projectmilestone_progress',
+		]);
+		$dataReader = $relatedListView->getRelationQuery()
+			->andWhere(['or', ['parentid' => 0], ['parentid' => null]])
+			->createCommand()->query();
+		while ($row = $dataReader->read()) {
+			$milestoneRecordModel = Vtiger_Record_Model::getInstanceById($row['id']);
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: ?

File: modules/Project/models/Module.php [hunk 12]
```diff
@@ -23,7 +24,158 @@ public function getSideBarLinks($linkParams)
...
+	 * @param int   $id
+	 * @param float $estimatedWorkTime
+	 *
+	 * @throws \App\Exceptions\AppException
+	 *
+	 * @return float
+	 */
+	public function calculateEstimatedWorkTime(int $id, float $estimatedWorkTime = 0): float
+	{
+		if (!App\Record::isExists($id)) {
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: ?

File: install/install_schema/data.sql [hunk 13]
```diff
@@ -4816,6 +4816,8 @@ insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_clas
 insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (63,'EntityBeforeSave','OSSTimeControl_TimeControl_Handler',1,'OSSTimeControl','',5,51);
 insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (64,'EntityBeforeSave','Vtiger_SocialMedia_Handler',1,'','',6,0);
 insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (65,'EntityBeforeDelete','Vtiger_SocialMedia_Handler',1,'','',6,0);
+insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (66,'EntityAfterSave','Project_ProjectHandler_Handler',1,'Project','',3,42);
+insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (67,'EntityChangeState','Project_ProjectHandler_Handler',1,'Project','',3,42);
```
Comment (rskrzypczak) [inline install/install_schema/data.sql]: ?

File: modules/Project/models/Module.php [hunk 14]
```diff
@@ -11,6 +12,66 @@
...
+	 * Get children by parent ID.
+	 *
+	 * @param int $id
+	 *
+	 * @return int[]
+	 */
+	protected static function getChildren(int $id): array
+	{
+		$queryGenerator = new \App\QueryGenerator('Project');
+		$queryGenerator->addNativeCondition(['parentid' => $id]);
```
Comment (rskrzypczak) [inline modules/Project/models/Module.php]: $queryGenerator->permissions = false;

File: install/install_schema/data.sql (hunk 13 above)
Comment (mariuszkrzaczkowski) [reply install/install_schema/data.sql]: ?

File: modules/Project/models/Module.php [hunk 15]
```diff
@@ -11,6 +12,91 @@
...
+	 * Calculate estimated work time.
+	 *
+	 * @param int   $id
+	 * @param float $estimatedWorkTime
+	 *
+	 * @throws \App\Exceptions\AppException
+	 *
+	 * @return float
+	 */
+	public static function calculateEstimatedWorkTime(int $id, float $estimatedWorkTime = 0): float
```
Comment (mariuszkrzaczkowski) [inline modules/Project/models/Module.php]: module > record

File: modules/Project/models/Module.php [hunk 16]
```diff
@@ -11,6 +12,68 @@
...
+	 *
+	 * @param int $id
+	 *
+	 * @return int[]
+	 */
+	protected static function getChildren(int $id): array
+	{
+		$queryGenerator = new \App\QueryGenerator('Project');
+		$queryGenerator->permissions = false;
+		$queryGenerator->addNativeCondition(['parentid' => $id]);
```
Comment (mariuszkrzaczkowski) [inline modules/Project/models/Module.php]: add support for archive records

File: modules/ProjectMilestone/models/Module.php [hunk 17]
```diff
@@ -10,6 +10,101 @@
...
+	 * @param int   $id
+	 * @param float $estimatedWorkTime
+	 *
+	 * @throws \App\Exceptions\AppException
+	 *
+	 * @return float
+	 */
+	public static function calculateEstimatedWorkTime(int $id, float $estimatedWorkTime = 0): float
+	{
+		$progressInHours = 0;
```
Comment (mariuszkrzaczkowski) [inline modules/ProjectMilestone/models/Module.php]: add a cache

**Pull request closed without merging**